Provide a standalone plugin manager implementation

weierophinney opened the pull request:
This patch allows zend-hydrator to be used without zend-servicemanager, making that package a truly optional dependency.

First, it adds `Zend\Hydrator\HydratorPluginManagerInterface`, which extends the PSR-11 `ContainerInterface`. This is done to ensure that any plugin manager implementation complies with PSR-11 (currently, `AbstractPluginManager` implementations from zend-servicemanager do not), as well as to allow type-hinting on that interface when a plugin manager is needed. The `DelegatingHydratorFactory` was updated to test against this interface instead of the `HydratorPluginManager`.

Next, the patch adds `Zend\Hydrator\StandaloneHydratorPluginManager`, which provides a minimal plugin manager that supports loading *only the shipped hydrator implementations*. A factory for creating the plugin manager is also provided via `Zend\Hydrator\StandaloneHydratorPluginManagerFactory`. 

`Zend\Hydrator\ConfigProvider` was updated to add the `StandaloneHydratorPluginManager` factory mapping, as well as to change the target of the `HydratorManager` alias based on the presence of the zend-servicemanager package.

A page on plugin managers was added, and includes an example of creating a custom plugin manager.

Fixes #72

Ocramius commented on src/StandaloneHydratorPluginManager.php:
```diff
@@ -0,0 +1,232 @@
+     *     alias to a previously resolved alias, a service mapped to a
+     *     factory, or a fully qualified class name.
+     */
+    public function addAlias(string $alias, string $target) : void
```
Do we need this sort of complexity in the base plugin manager? Assuming we only expose the PSR-11 API when falling back to this container, we can probably leave out most of it, no?

weierophinney replied:
I think having something that doesn't allow adding services makes it pretty much useless, as the point is to allow pulling either specific types, or specifically configured instances, and the latter two are likely to be user-generated. If we don't have _some_ capabilities for adding to the manager, users are then forced to create their own implementation as soon as they need to use a custom hydrator, or want one of the shipped hydrators with specific customisations (strategies, filters, etc.).

We _could_ limit it to only factories, but that can potentially lead to a lot of duplicated service definitions. I felt limiting to aliases, invokables (via aliases), and factories was a reasonable compromise in this situation.

Ocramius replied:
> I think having something that doesn't allow adding services makes it pretty much useless

It's a fallback DIC for setups where a configurable DIC is not wanted. This is the scenario where users just want the hydrators internals to work without worrying too much about them.

I'd even just hardcode the map of services in an inlined `get` and `has` in a minimal PSR-11 container implementation here...

weierophinney replied:
Fair enough; if they want to do more, they can install zend-servicemanager, wire them into the container they use normally, or create their own container.

I'll update the PR this morning to do that, and document the approaches (including an example of how to write a custom container).

weierophinney commented:
@Ocramius I've updated the patch as requested. The `StandaloneHydratorPluginManager` now uses (mostly) hard-coded lookups (I'm allowing casing variations for unqualified names). I've also added a section to the plugin managers documentation detailing how you might write a custom version.

Ocramius commented on docs/book/v3/plugin-managers.md:
```diff
@@ -0,0 +1,197 @@
+    }
+}
+
+// in config/autoload/hydrators.global.php or similar:
```
Separate in a different quoted code block?

Ocramius commented on docs/book/v3/plugin-managers.md:
```diff
@@ -0,0 +1,197 @@
+use Zend\Hydrator\HydratorPluginManagerInterface;
+use Zend\Hydrator\StandaloneHydratorPluginManager;
+
+class CustomHydratorPluginManager implements
```
I think this section goes too deep in detail: describing what interface is needed should be sufficient, no?

weierophinney replied:
It may be too much detail for you or me, but having a concrete example to look at is really useful for junior and medior developers. I want to demonstrate a complete solution - even if it's something they would need to alter for their specific needs.

Ocramius replied:
This is code sitting in the docs, hidden in a block in a very edge case area: I'd simplify it to the bone to prevent that.

Ocramius commented on docs/book/v3/plugin-managers.md:
```diff
@@ -0,0 +1,197 @@
+use Zend\Hydrator\StandaloneHydratorPluginManager;
+
+class CustomHydratorPluginManager implements
+    ContainerInterface,
```
`HydratorPluginManagerInterface` already contains `ContainerInterface`

Ocramius commented on src/HydratorPluginManagerInterface.php:
```diff
@@ -0,0 +1,16 @@
+
+use Psr\Container\ContainerInterface;
+
+interface HydratorPluginManagerInterface extends ContainerInterface
```
Should we add `@method` which describes `get()` to be returning a hydrator?

Ocramius commented on src/StandaloneHydratorPluginManager.php:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * @var callable Invokable factory for hydrators without dedicated factories.
+     */
+    private $invokableFactory;
```
This property is only ever used in the constructor: not needed

Ocramius commented on src/StandaloneHydratorPluginManager.php:
```diff
@@ -0,0 +1,138 @@
+            return $this->aliases[$name];
+        }
+
+        $alias = strtolower($name);
```
Variable redundant

Ocramius commented on src/StandaloneHydratorPluginManager.php:
```diff
@@ -0,0 +1,138 @@
+     * Next it checks if the $name is registered verbatim in $aliases; if so,
+     * it returns the target of the alias.
+     *
+     * finally, it does a strtolower() on it and looks to see if it exists
```
This is a description of the implementation: since we are in `private` scope, most of the docblock, whilst extensive and descriptive, is kinda not needed

Ocramius commented on src/StandaloneHydratorPluginManager.php:
```diff
@@ -0,0 +1,138 @@
+ * implementation.
+ */
+final class StandaloneHydratorPluginManager implements
+    ContainerInterface,
```
Redundant declaration

Ocramius commented on src/StandaloneHydratorPluginManagerFactory.php:
```diff
@@ -0,0 +1,20 @@
+
+use Psr\Container\ContainerInterface;
+
+final class StandaloneHydratorPluginManagerFactory
```
Can probably be simplified with an invokable factory?

weierophinney replied:
That would require using zend-servicemanager... and the whole point of having the standalone version is when you do not want to use that implementation.

Ocramius replied:
Doesn't the `ConfigProvider` already retrieve a naming structure that assumes `zend-servicemanager`? 

https://github.com/zendframework/zend-hydrator/pull/87/files#diff-e97918d312824c7b64da5ef255c18258

weierophinney replied:
Yes and no!

We co-opted that for containers compatible with Expressive, and document it there as well: https://docs.zendframework.com/zend-expressive/v3/features/container/config/

So if you're using any of the 5 or so other compatible implementations, this will continue to work. It will also work if you're using the component in another context entirely, which is the important bit.

Ocramius replied:
:+1:

malukenho commented on test/StandaloneHydratorPluginManagerTest.php:
```diff
@@ -0,0 +1,116 @@
+
+class StandaloneHydratorPluginManagerTest extends TestCase
+{
+    public function setUp()
```
IMHO we should keep the visibility from the parent class, so it could be `protected` instead.

weierophinney replied:
We use public visibility pretty consistently everywhere for the setUp and tearDown methods. I'm not going to change it here.
